### DISCUSSION TITLE
Kontrola branchy git w azure devops

### DISCUSSION THREAD

**Discussion by maciej-skorupka** [AzureCommunityPL/group#317] [Pomoc]
Hej,

Czego szukam? Sposobu na kontrolę powstawania branchy w git devops. Rzeczy w stylu https://docs.microsoft.com/pl-pl/azure/devops/repos/git/require-branch-folders?view=azure-devops&tabs=browser ale bez użycia tf.exe wspomnianego tu, bo wydaje się on jest częścią nieużywanego przeze mnie visual studio (używam vscode). Nie znalazłem na razie niczego co daje podobne ficzery i szukam inspiracji:)

**Comment by kamil-mrzyglod** [comment]
Problem z tym, że nie bardzo masz taką możliwość(a przynajmniej ja nie kojarzę natywnego mechanizmu). Dla mnie najłatwiejszym sposobem(zakładając, że masz jakąkolwiek kontrolę nad środowiskiem deweloperskim) było zawsze użycie git hooks. Plus jest taki, że masz wtedy uniwersalne rozwiązanie działające z każdym narzędziem.

**Reply by mgrabarz** [reply]
W teorii, jeśli zewnętrzne narzędzia coś potrafią zrobić, to ty pewnie też przez API.
Azure DevoOps ma dość rozbudowane API, do tego webhooki, o których pisal Kamil. Potem trochę kodu, logiki w kodzie i może osiągniesz oczekiwany efekt.

W Github jest nieco łatwiej, bo masz Actions, lub gotowe narzędzia jak Probot.

**Reply by JanuszNowak** [reply]
Zależnie od potrzeby możliwości 
- Client-side hooks
- Server-side hooks
- Pull Request Status Server
- Build as Trigger to check name
- Uprawienie na poziomie folderów przez tf.exe lub przez Gui ADO, pewnie tez przez Rest API, zablokuje to wypchania danej gałęzi jeżeli nie ma sie uprawnień
- Zewnętrzną mała apka która chodzi co chwile i skanuje 
- Pull Request Status Server - chyba najlepszym rozwiązaniem bo masz pełno kontrole co jest valid a co nie
